### PR TITLE
Move BOLT from `bootstrap` to `opt-dist`

### PR DESCRIPTION
**Pull request by Kobzol**
Currently, we use BOLT to optimize LLVM for x64 Linux. The BOLT instrumentation and optimization step is implemented in `bootstrap`, but it was always quite hacky, because BOLT works quite differently than PGO. Rather than building an instrumented artifact, it takes an already built artifact and instruments it in-place. This is not a good fit for the bootstrap caching mechanism, and it meant that we had to run BOLT "on-the-fly" when packaging LLVM artifacts into the created sysroot.

The BOLT code was also really only used by the PGO script (now called `opt-dist`) and nothing else, so it was quite hardcoded for this one single usage. And even if someone wanted to use the `--llvm-bolt-profile-[use/generate]` bootstrap flags outside of the PGO script, they would also need to implement profile gathering, as this is not performed by bootstrap anyway.

I think that it could be more practical to move the BOLT logic to the `opt-dist` tool instead. This simplifies bootstrap, removes unneeded implementation of BOLT caching (we will now do it exactly once - no need to check if it has been performed already when bootstrap copies `libLLVM.so` around multiple times) and removes two BOLT-specific bootstrap flags, and also one special case for building LLVM (instead I pass the linker flags with `--set llvm.ldflags` from `opt-dist` now).

There are also a few disadvantages to this new approach:
- We have to guess/find the path to the built `libLLVM.so` file (but currently this is quite easy, it's just in `stage2/lib`).
- We have to provide the BOLT profile externally to bootstrap, so that it is packaged into the reproducible artifacts archive. Doesn't seem like a big deal to me.
- We are depending on some inner behavior of boostrap in `opt-dist` (namely, that `libLLVM.so` is hardlinked). But we do that for many other things in the `opt-dist` tool anyway, it's tied quite closely to bootstrap.

I would like to go back to my attempts to also use BOLT for `librustc_driver.so`, and I think that it might be a bit simpler if I also do it from the `opt-dist` tool, so this is the first step towards that.

Anyway, let me know what you think about this. It's just a refactoring in a way, no big deal.

r? bootstrap

### CODE REVIEW

File: src/tools/opt-dist/src/utils/io.rs [hunk 1]
```diff
@@ -46,3 +53,28 @@ pub fn unpack_archive(path: &Utf8Path, dest_dir: &Utf8Path) -> anyhow::Result<()
...
+    directory: &Utf8Path,
+    prefix: &str,
+    suffix: &str,
+) -> anyhow::Result<Utf8PathBuf> {
+    let mut files = glob::glob(&format!("{directory}/{prefix}*{suffix}"))?
+        .into_iter()
+        .collect::<Result<Vec<_>, _>>()?;
+    match files.pop() {
+        Some(file) => {
+            if !files.is_empty() {
```
Comment (ozkanonur) [inline src/tools/opt-dist/src/utils/io.rs:68]: Not hard feelings: Would be easier to read this with less nested blocks by `Some(file) if !files.is_empty()`

**Comment by ozkanonur** [comment]
I also want to see Mark's review on this.

r? @Mark-Simulacrum

File: src/tools/opt-dist/src/exec.rs [hunk 2]
```diff
@@ -144,13 +144,13 @@ impl Bootstrap {
         self
     }
 
-    pub fn llvm_bolt_instrument(mut self) -> Self {
-        self.cmd = self.cmd.arg("--llvm-bolt-profile-generate");
+    pub fn with_llvm_bolt_ldflags(mut self) -> Self {
+        self.cmd = self.cmd.arg("--set").arg("llvm.ldflags=-Wl,-q");
```
Comment (lqd) [inline src/tools/opt-dist/src/exec.rs:148]: will this append a flag or replace the possibly existing `ldflags` in the config ?

Comment (Kobzol) [reply src/tools/opt-dist/src/exec.rs:148]: It appends: https://github.com/rust-lang/rust/blob/3e03a48c18c4465380dca49a41add1609c09d2d8/src/bootstrap/llvm.rs#L716.

File: src/tools/opt-dist/src/utils/io.rs (hunk 1 above)
Comment (lqd) [reply src/tools/opt-dist/src/utils/io.rs:68]: Or check for the length !=1 errors before the match, removing the need to push back the thing you popped to restore the original list.

**Comment by Kobzol** [comment]
Rebased to incorporate https://github.com/rust-lang/rust/pull/114141.

**Comment by Kobzol** [comment]
Gentle ping @Mark-Simulacrum, just to make sure that this isn't forgotten.

**Pull request merged**